src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.pi.lex
+import data.finsupp.order
+
+/-!
+# Lexicographic order on finitely supported functions
+
+This file defines the lexicographic order on `finsupp`.
+-/
+
+namespace finsupp
+
+variables {α N : Type*} [has_zero N]
+
+/-- The lexicographic relation on `α →₀ N`, where `α` is ordered by `r`,
+  and `N` is ordered by `s`. -/
+protected def lex (r : α → α → Prop) (s : N → N → Prop) (x y : α →₀ N) : Prop :=
+pi.lex r s x y
+
+instance [has_lt α] [has_lt N] : has_lt (lex (α →₀ N)) := ⟨finsupp.lex (<) (<)⟩
+
+instance lex.is_strict_order [linear_order α] [partial_order N] :
+  is_strict_order (lex (α →₀ N)) (<) :=
+{ irrefl := λ a ⟨_, _, h⟩, lt_irrefl _ h,
+  trans := begin
+    rintro a b c ⟨N₁, lt_N₁, a_lt_b⟩ ⟨N₂, lt_N₂, b_lt_c⟩,
+    rcases lt_trichotomy N₁ N₂ with (H|rfl|H),
+    exacts [⟨N₁, λ j hj, (lt_N₁ _ hj).trans (lt_N₂ _ $ hj.trans H), lt_N₂ _ H ▸ a_lt_b⟩,
+      ⟨N₁, λ j hj, (lt_N₁ _ hj).trans (lt_N₂ _ hj), a_lt_b.trans b_lt_c⟩,
+      ⟨N₂, λ j hj, (lt_N₁ _ (hj.trans H)).trans (lt_N₂ _ hj), (lt_N₁ _ H).symm ▸ b_lt_c⟩]
+  end }
+
+lemma filter_ne_eq_empty_iff [decidable_eq α] [decidable_eq N] {f g : α →₀ N} :
+  (f.support ∪ g.support).filter (λ a, f a ≠ g a) = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp⟩,
+  ext a,
+  contrapose h,
+  refine finset.ne_empty_of_mem (_ : a ∈ _),
+  simp only [ne.def, finset.mem_filter, finset.mem_union, finsupp.mem_support_iff],
+  exact ⟨not_and_distrib.1 $ mt (and_imp.2 eq.substr) (ne.symm h), h⟩,
+end
+
+variables [linear_order α]
+
+/--  The partial order on `finsupp`s obtained by the lexicographic ordering.
+`finsupp.lex.linear_order` is the proof that this partial order is in fact linear. -/
+instance lex.partial_order [partial_order N] : partial_order (lex (α →₀ N)) :=
+partial_order.lift _ (@id (function.injective (_ : lex (α →₀ N) → lex (α → N)))
+  finsupp.coe_fn_injective)
```

Comment:
Shorter and with better definition reduction:
```suggestion
partial_order.lift (λ x, to_lex ⇑(of_lex x)) finsupp.coe_fn_injective
```

Reply:
[finsupp.coe_fn_injective](https://leanprover-community.github.io/mathlib_docs/data/finsupp/basic.html#finsupp.coe_fn_injective) is deprecated. Can you use [fun_like.coe_injective](https://leanprover-community.github.io/mathlib_docs/data/fun_like/basic.html#fun_like.coe_injective) instead? I argued [in another PR](https://github.com/leanprover-community/mathlib/pull/15889#discussion_r943050890) that it shouldn't be deprecated, but I was able to remove its use anyway.

Reply:
I've found this version harder to work with: can you find a better end for [this part](https://github.com/leanprover-community/mathlib/blob/f75518db61cea68980db1725a1ef29c8f6d5f79e/src/data/finsupp/lex.lean#L67-L69), where `simp [js]` used to work?  (The same happened at L75-L77.

Before "that" happens, the goal `dsimp`s to
```
⇑(⇑of_lex g) j = ⇑(⇑of_lex f) j
```
and I can't get Lean to realize that it should write it as
```
⇑g j = ⇑f j
```
but accepts that the two are defeq if I spoon-feed it the expressions.

Reply:
Btw, adding
```lean
lemma of_lex_apply [partial_order N] {f : lex (α →₀ N)} {j : α} : (of_lex f) j = f.to_fun j := rfl
```
does not help.

Reply:
It is not the correct approach. `of_lex x` is def-eq to `x` but this breaks API boundaries. You're only ever supposed to use `to_lex` and `of_lex` to cast between these.

Reply:
Maybe I'm missing something, but when below I use `exact _` to prove the goal I am again "crossing API boundaries", right?

What is a good way of proceeding?

Reply:
> Before "that" happens, the goal `dsimp`s to
> 
> ```lean
> ⇑(⇑of_lex f) j = ⇑(⇑of_lex g) j
> ```
> 
> and I can't get Lean to realize that it should write it as
> 
> ```lean
> ⇑f j = ⇑g j
> ```
> 
> but accepts that the two are defeq if I spoon-feed it the expressions.

That second version doesn't make sense, assuming `f` and `g` are of type `lex _`

Reply:
But `f g` have type `α →₀ N`.  Maybe I should put a type ascription when the first `lam` introduces them.

Reply:
Ok, with the explicit type-ascription and the forall\2, `simp` now works!  Thanks!

Reply:
The type ascription is at best redundant or at worst a bad idea